fe2-android/app/src/main/java/com/github/dedis/student20_pop/model/data/LAORepository.java

```diff
@@ -55,504 +53,593 @@
+            MessageGeneral msg = messageById.get(messageId);
+            msg.getWitnessSignatures().add(new PublicKeySignaturePair(senderPkBuf, signatureBuf));
 
-    Optional<RollCall> rollCallOptional = lao.getRollCall(closes);
-    if (!rollCallOptional.isPresent()) {
-      return true;
+            Lao lao = laoById.get(channel).getLao();
+            if (lao == null) {
+                Log.d(TAG, "failed to retrieve the lao with channel " + channel);
+                return false;
+            }
+            // Update WitnessMessage of the corresponding lao
+            if (!updateWitnessMessage(lao, messageId, senderPk)) {
+                return false;
+            }
+
+            Set<PendingUpdate> pendingUpdates = lao.getPendingUpdates();
+            if (pendingUpdates.contains(messageId)) {
+                // We're waiting to collect signatures for this one
+
+                // Let's check if we have enough signatures
+                Set<String> signaturesCollectedSoFar =
+                        msg.getWitnessSignatures().stream()
+                                .map(ob -> Base64.getUrlEncoder().encodeToString(ob.getWitness()))
+                                .collect(Collectors.toSet());
+                if (lao.getWitnesses().equals(signaturesCollectedSoFar)) {
+
+                    // We send a state lao if we are the organizer
+                    sendStateLao(lao, msg, messageId, channel);
+
+                }
+            }
+
+            return false;
+        }
+
+        return true;
     }
 
-    RollCall rollCall = rollCallOptional.get();
-    rollCall.setEnd(closeRollCall.getClosedAt());
-    rollCall.setId(updateId);
-    rollCall.getAttendees().addAll(closeRollCall.getAttendees());
-    rollCall.setState(EventState.CLOSED);
-
-    lao.updateRollCall(closes, rollCall);
-    return false;
-  }
-
-  private boolean handleWitnessMessage(String channel, String senderPk, WitnessMessage message) {
-    String messageId = message.getMessageId();
-    String signature = message.getSignature();
-
-    byte[] senderPkBuf = Base64.getUrlDecoder().decode(senderPk);
-    byte[] signatureBuf = Base64.getUrlDecoder().decode(signature);
-
-    // Verify signature
-    try {
-      PublicKeyVerify verifier = new Ed25519Verify(senderPkBuf);
-      verifier.verify(signatureBuf, Base64.getUrlDecoder().decode(messageId));
-    } catch (GeneralSecurityException e) {
-      Log.d(TAG, "failed to verify witness signature " + e.getMessage());
-      return false;
+    /**
+     * Helper method to verify signature of a message
+     *
+     * @param messageId    Base 64 URL encode Id of the message to sign
+     * @param senderPkBuf  Base 64 decoded public key of the signer
+     * @param signatureBuf Base 64 URL decoded signature of the signer
+     * @return false if there was a problem signing the message
+     */
+
+    private boolean verifySignature(String messageId, byte[] senderPkBuf, byte[] signatureBuf) {
```

Comment:
Same as the `generateSignature` method, these are utility methods to sign and verify that correspond in a separate class in the package utility/security